tracer/src/Datadog.Trace/ClrProfiler/CallTarget/Handlers/IntegrationOptions.cs

```diff
@@ -30,13 +32,94 @@ internal static void LogException(Exception exception, string message = null)
+                if (_integrationId.Value is not null)
+                {
+                    Tracer.Instance.TracerManager.Telemetry.IntegrationDisabledDueToError(_integrationId.Value.Value, nameof(CallTargetInvokerException));
+                }
+
                 _disableIntegration = true;
             }
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void RecordTelemetry()
+        {
+            if (_integrationId.Value is not null)
+            {
+                Tracer.Instance.TracerManager.Telemetry.IntegrationRunning(_integrationId.Value.Value);
+            }
+        }
+
+        // internal for testing
+        internal static IntegrationId? GetIntegrationId(Type integrationType, Type targetType)
+        {
+            try
+            {
+                var attributes = integrationType.GetCustomAttributes(typeof(InstrumentMethodAttribute), inherit: true);
+                if (attributes.Length > 0)
+                {
+                    var name = ((InstrumentMethodAttribute)attributes[0]).IntegrationName;
+                    if (string.IsNullOrEmpty(name))
+                    {
+                        // shouldn't happen?
+                        return null;
+                    }
+
+                    return IntegrationRegistry.TryGetIntegrationId(name, out var integrationId) ? integrationId : null;
+                }
+
+                // probably assembly attribute
+                var assemblyAttributes = integrationType.Assembly.GetCustomAttributes(typeof(InstrumentMethodAttribute), inherit: true);
+                if (assemblyAttributes.Length == 0)
+                {
+                    // can't work out which integration this is, shouldn't happen?
+                    return null;
+                }
+
+                var targetTypeName = targetType.FullName;
+                var targetAssemblyName = targetType.Assembly.GetName().Name;
+
+                foreach (InstrumentMethodAttribute assemblyAttribute in assemblyAttributes)
+                {
+                    if (assemblyAttribute.TypeName == targetTypeName)
+                    {
+                        if (assemblyAttribute.AssemblyNames is not null)
+                        {
+                            foreach (var assemblyName in assemblyAttribute.AssemblyNames)
+                            {
+                                if (assemblyName == targetAssemblyName)
+                                {
+                                    return IntegrationRegistry.TryGetIntegrationId(assemblyAttribute.IntegrationName, out var integrationId) ? integrationId : null;
+                                }
+                            }
+                        }
+
+                        if (assemblyAttribute.AssemblyName is not null
+                         && assemblyAttribute.AssemblyName == targetAssemblyName)
```

Comment:
Likewise, we also need the fix here:
```suggestion
                         && assemblyAttribute.AssemblyName == targetAssemblyName
                         && assemblyAttribute.CallTargetType.FullName == integrationType.FullName)
```